jsonrpcservice/__init__.py

```diff
@@ -0,0 +1,259 @@
+        self._prepare()
+        return self._kwargs
+
+    def is_notification(self):
+        self._prepare()
+        return self._notification
+
+    def _prepare(self):
+        if self._prepared:
+            return
+
+        self._prepared = True
+
+        data = self.parsed_request()
+
+        if not isinstance(data, dict):
+            raise InvalidRequest('Expected an object as request body. Batch requests are not supported.')
+
+        # Remember `id` as the first thing. `id` can be ommited for notifications.
+        self._id = data.get('id')
+
+        if 'jsonrpc' not in data:
+            raise InvalidRequest('Missing "jsonrpc" member in request.')
+
+        if data['jsonrpc'] != "2.0":
+            raise InvalidRequest('Only JSON-RPC version "2.0" is supported.')
+
+        if self._id is None:
+            # Flag as notification only if client specified protocol correctly and thus should know he can't expect
+            # response if id is not set.
+            self._notification = True
+
+        if 'method' not in data:
+            raise InvalidRequest('Missing "method" member in request.')
+
+        if not isinstance(data['method'], str):
+            raise InvalidRequest('"method" of request must be a string.')
+
+        self._method = data['method']
+
+        if 'params' in data:
+            if isinstance(data['params'], list):
+                self._args.extend(data['params'])
+            elif isinstance(data['params'], dict):
+                self._kwargs.update(data['params'])
+            else:
+                raise InvalidRequest('"params" of request must be either object or array if present.')
+
+
+class Service(object):
+    def __init__(self):
+        self._methods = {}
+
+    def method(self, method):
+        if callable(method):
+            self.add_method(method.__name__, method)
+
+            return method
+        else:
+            def wrapper(func):
+                self.add_method(method, func)
+                return func
+
+            return wrapper
+
+    def add_method(self, method_name, func):
+        if method_name in self._methods:
+            raise ValueError('Method "{}" already registered.'.format(method_name))
+
+        self._methods[method_name] = func
+
+    def add_methods(self, methods, prefix=''):
+        if isinstance(methods, (list, tuple)):
+            methods = {method.__name__: method for method in methods}
+        elif not isinstance(methods, dict):
+            methods = {method: getattr(methods, method) for method in dir(methods) if not method.startswith('_')}
+
+        for name, method in methods.items():
+            if callable(method):
+                self.add_method(prefix + name, method)
+
+    def get_methods(self):
+        return self._methods.copy()
+
+    def handle_request(self, request):
+        response = None
+        try:
+            method_name, args, kwargs = self.before_call(request, request.method(), request.args(), request.kwargs())
```

Comment:
Why do we need those `before_call` and `after_call` methods? From what can I see, all of the extensibility use cases can be handled in middleware-style, by overriding the `call_method`. Am I missing something?

Reply:
Convenience, semantics. `after_call` is always called even if `call_method` raises error.

Why override core method, when you can override just a hook?

Reply:
You can achieve the same behaviour by wrapping `super().call_method(...)` with `try-except`.

Who said `call_method` is core? What prevents us to document it and allow others to override it? From the other point of view – why introduce two more methods to public interface, when there can be only one method which can handle all of the cases.

My main motivation comes from PEP20:
> There should be one-- and preferably only one --obvious way to do it.

Reply:
You can achieve the same by writing it yourself. There never can be only one way. 

I want to give user of this library the right tools to do what they want to do. Giving them hooks for specific purpose is not evil, it guides them into doing things safer way if possible. It helps extendability and reusability. It is simple as fuck, the negative consequences of giving them this little simple tool are close to zero.

If you don't want to use it, don't.

And please learn to distinguish between simplified recommendations for noobs and an unbreakable dogma.